### PR TITLE
Blog on DDEV golang debugging

### PR DESCRIPTION
**Pull request by rfay**
This is a summary of last Wednesday's Contributor Training on DDEV Golang Debugging. 

Suggestions and review welcome!

Rendered version at https://20240517-go-debugging.ddev-com-front-end.pages.dev/blog/golang-debugging

### CODE REVIEW

File: src/content/blog/golang-debugging.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
+---
+title: "Debugging Golang (Go) Applications"
+pubDate: 2024-05-17
+# modifiedDate: 2024-04-23
+summary: Debugging Golang applications using Goland or Vscode
```
Comment (stasadev) [inline src/content/blog/golang-debugging.md:5]:
```suggestion
summary: Debugging Golang applications using GoLand or VS Code
```

File: src/content/blog/golang-debugging.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+It's not hard to work with DDEV's Golang code, but you definitely need a development environment and the know-how to do step-debugging. Here's our [Contributor Training](contributor-training.md) showing some of the nuances with DDEV, followed by a short summary of the details.
+
+<iframe width="560" height="315" src="https://www.youtube.com/embed/E-AEzC1p76E?si=XYP23HYcxgqiJ2_M" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>
+
+First, you need some of the basics:
+
+* Install `go`, `make`, and `delve` (needed for Visual Studio Code only)
+  * On macOS: `brew install golang delve` (`make` is already provided by Apple's Xcode command-line tools)
+  * WSL2/Linux: `sudo snap install --classic go && sudo apt update && sudo apt install -y build-essential clang ccache`
+* Install your IDE, Goland or vscode. On Windows/WSL2 most people install each of these on the Windows side, although they can be run inside WSL2 using the Linux versions.
```
Comment (stasadev) [inline src/content/blog/golang-debugging.md:24]:
```suggestion
* Install your IDE, [GoLand](https://www.jetbrains.com/go/) or [VS Code](https://code.visualstudio.com/). On Windows/WSL2 most people install each of these on the Windows side, although they can be run inside WSL2 using the Linux versions.
```

File: src/content/blog/golang-debugging.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+* For Goland, nothing further is required, but on vscode you'll need to install the [golang go](https://marketplace.visualstudio.com/items?itemName=golang.go) extension.
```
Comment (stasadev) [inline src/content/blog/golang-debugging.md:25]:
```suggestion
* For GoLand, nothing further is required, but on VS Code you'll need to install the [golang go](https://marketplace.visualstudio.com/items?itemName=golang.go) extension.
```

File: src/content/blog/golang-debugging.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+Once you've have your IDE working, you can start debugging.
+
+## Goland Debugging
+
+### Debug a simple command like `ddev list`
+
+* Open `main.go` in Goland
+* Right-click on `main.go` in the left pane (`cmd/ddev/main.go`) and click `Run`
+* Goland will create a "Run/Debug Configuration" named `go build main.go`
```
Comment (stasadev) [inline src/content/blog/golang-debugging.md:35]:
```suggestion
* GoLand will create a "Run/Debug Configuration" named `go build main.go`
```

File: src/content/blog/golang-debugging.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+* Click the arrowhead (run) or "bug" (debug) symbol next to the test
+* Or right-click on a `*_test.go` file to run all the tests in the file.
+
+### Debug a test in the `cmd` directory
+
+In the `cmd` directory most tests actually *run* the `ddev` binary, so you'll need to make sure that the `ddev` in the PATH is the one that represents the code you're testing. For example, `TestDebugMigrateDatabase` in `cmd\ddev\cmd\debug-migrate-database_test.go` actually executes `ddev debug migrate-database`, so if you don't have the right `ddev` in your PATH, your test is not valid.
+
+Otherwise, everything is the same. Just click the arrowhead or the "bug" beside a test and let it run.
+
+## Visual Studio Code (vscode) Debugging
```
Comment (stasadev) [inline src/content/blog/golang-debugging.md:59]:
```suggestion
## Visual Studio Code (VS Code) Debugging
```

File: src/content/blog/golang-debugging.md [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+Vscode debugging is a bit more tweaky. It seems that for commands, you need to configure what you want to use in the launch.json. DDEV code comes with a built-in launch.json, which you can copy-and-paste as needed. For example:
```
Comment (stasadev) [inline src/content/blog/golang-debugging.md:61]:
```suggestion
VS Code debugging is a bit more tweaky. It seems that for commands, you need to configure what you want to use in the launch.json. DDEV code comes with a built-in launch.json, which you can copy-and-paste as needed. For example:
```

File: src/content/blog/golang-debugging.md [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+In DDEV, the tests in the `pkg` directory have no dependencies, you can just run them.
+
+* Find a test you want to run.
+* Click the arrowhead (run) or "bug" (debug) symbol next to the test
+* Or right-click on a `*_test.go` file to run all the tests in the file.
+
+### Debug a test in the `cmd` directory
+
+In the `cmd` directory most tests actually *run* the `ddev` binary, so you'll need to make sure that the `ddev` in the PATH is the one that represents the code you're testing. For example, `TestDebugMigrateDatabase` in `cmd\ddev\cmd\debug-migrate-database_test.go` actually executes `ddev debug migrate-database`, so if you don't have the right `ddev` in your PATH, your test is not valid.
```
Comment (stasadev) [inline src/content/blog/golang-debugging.md:55]:
```suggestion
In the `cmd` directory most tests actually *run* the `ddev` binary, so you'll need to make sure that the `ddev` in the `$PATH` is the one that represents the code you're testing. For example, `TestDebugMigrateDatabase` in `cmd\ddev\cmd\debug-migrate-database_test.go` actually executes `ddev debug migrate-database`, so if you don't have the right `ddev` in your `$PATH`, your test is not valid.
```

File: src/content/blog/golang-debugging.md [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+* `ddev start` will happen, and it will stop at the line where you set the breakpoint.
+* Classic debugger behavior is available, either with icons or function keys, and you can examine variable values as you go.
+
+### Running a test
+
+Running tests is almost exactly the same as in Goland. Right-click the arrow next to the test function and choose "Run" or "Debug". The same caveats apply for `cmd` tests, you need `ddev` built from the same code in your `$PATH`.
+
+## Working in Gitpod
+
+Every DDEV PR and the main page has a launcher for Gitpod. Gitpod is just a Linux-based vscode development environment in the cloud, in a browser. You can do everything you would want to do with it that you could do with vscode locally. If you click the "Launch in Gitpod" button on any PR, that PR will be set up for you automatically in Gitpod, with all vscode extensions already loaded, and with DDEV already compiled.
```
Comment (stasadev) [inline src/content/blog/golang-debugging.md:79]:
```suggestion
Every DDEV PR and the main page has a launcher for Gitpod. Gitpod is just a Linux-based VS Code development environment in the cloud, in a browser. You can do everything you would want to do with it that you could do with VS Code locally. If you click the "Launch in Gitpod" button on any PR, that PR will be set up for you automatically in Gitpod, with all VS Code extensions already loaded, and with DDEV already compiled.
```

File: src/content/blog/golang-debugging.md [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+* Install your IDE, Goland or vscode. On Windows/WSL2 most people install each of these on the Windows side, although they can be run inside WSL2 using the Linux versions.
+* For Goland, nothing further is required, but on vscode you'll need to install the [golang go](https://marketplace.visualstudio.com/items?itemName=golang.go) extension.
+
+Once you've have your IDE working, you can start debugging.
+
+## Goland Debugging
+
+### Debug a simple command like `ddev list`
+
+* Open `main.go` in Goland
```
Comment (stasadev) [inline src/content/blog/golang-debugging.md:33]:
```suggestion
* Open `main.go` in GoLand
```

File: src/content/blog/golang-debugging.md [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+* Set a breakpoint in `cmd/ddev/cmd/start.go`
+* Click the triangle/bug icon at the left to "Run and Debug"
+* Select "debug ddev start"
+* Click the green arrowhead to the left of "debug ddev start"
+* `ddev start` will happen, and it will stop at the line where you set the breakpoint.
+* Classic debugger behavior is available, either with icons or function keys, and you can examine variable values as you go.
+
+### Running a test
+
+Running tests is almost exactly the same as in Goland. Right-click the arrow next to the test function and choose "Run" or "Debug". The same caveats apply for `cmd` tests, you need `ddev` built from the same code in your `$PATH`.
```
Comment (stasadev) [inline src/content/blog/golang-debugging.md:75]:
```suggestion
Running tests is almost exactly the same as in GoLand. Right-click the arrow next to the test function and choose "Run" or "Debug". The same caveats apply for `cmd` tests, you need `ddev` built from the same code in your `$PATH`.
```

**Review by stasadev — APPROVED** [review]
Thank you! I only have small suggestions for IDE naming.

**Comment by rfay** [comment]
@mandrasch - you're in the video thumbnail here. Is that OK? (Your review in general is welcome also)

**Comment by mandrasch** [comment]
> @mandrasch - you're in the video thumbnail here. Is that OK? (Your review in general is welcome also)

I saw it briefly on YouTube already, funny snapshot 😄 

If it is not too much hassle, please change it - I don't really feel like being the face of Go debugging (yet) 😄

Blog post is a great summary from my perspective, awesome work to document it afterwards 👍

**Comment by rfay** [comment]
Done, thanks.

**Pull request merged**